// react-redux-events/lib/actions.js
'use strict';

exports.__esModule = true;
exports.setLogLevel = exports.removeAllListenersForEvent = exports.removeAllListenersForContext = exports.removeEventListener = exports.addEventListener = exports.dispatchEvent = undefined;

var _constants = require('./constants');

var dispatchEvent = exports.dispatchEvent = function dispatchEvent(event) {
  return {
    type: _constants.DISPATCH_EVENT,
    event: event
  };
};

var addEventListener = exports.addEventListener = function addEventListener(name, context, handler) {
  var priority = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  return {
    type: _constants.ADD_EVENT_LISTENER,
    name: name,
    context: context,
    handler: handler,
    priority: priority
  };
};

var removeEventListener = exports.removeEventListener = function removeEventListener(name, context, handler) {
  return {
    type: _constants.REMOVE_EVENT_LISTENER,
    context: context,
    name: name,
    handler: handler
  };
};

var removeAllListenersForContext = exports.removeAllListenersForContext = function removeAllListenersForContext(context) {
  return {
    type: _constants.REMOVE_ALL_LISTENERS_FOR_CONTEXT,
    context: context
  };
};

var removeAllListenersForEvent = exports.removeAllListenersForEvent = function removeAllListenersForEvent(name) {
  return {
    type: _constants.REMOVE_ALL_LISTENERS_FOR_EVENT,
    name: name
  };
};

var setLogLevel = exports.setLogLevel = function setLogLevel(level) {
  return {
    type: _constants.SET_LOGLEVEL,
    level: level
  };
};
