// react-redux-events/es/actions.js
import { ADD_EVENT_LISTENER, REMOVE_EVENT_LISTENER, DISPATCH_EVENT, SET_LOGLEVEL, REMOVE_ALL_LISTENERS_FOR_EVENT, REMOVE_ALL_LISTENERS_FOR_CONTEXT } from './constants';

export var dispatchEvent = function dispatchEvent(event) {
  return {
    type: DISPATCH_EVENT,
    event: event
  };
};

export var addEventListener = function addEventListener(name, context, handler) {
  var priority = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;
  return {
    type: ADD_EVENT_LISTENER,
    name: name,
    context: context,
    handler: handler,
    priority: priority
  };
};

export var removeEventListener = function removeEventListener(name, context, handler) {
  return {
    type: REMOVE_EVENT_LISTENER,
    context: context,
    name: name,
    handler: handler
  };
};

export var removeAllListenersForContext = function removeAllListenersForContext(context) {
  return {
    type: REMOVE_ALL_LISTENERS_FOR_CONTEXT,
    context: context
  };
};

export var removeAllListenersForEvent = function removeAllListenersForEvent(name) {
  return {
    type: REMOVE_ALL_LISTENERS_FOR_EVENT,
    name: name
  };
};

export var setLogLevel = function setLogLevel(level) {
  return {
    type: SET_LOGLEVEL,
    level: level
  };
};
